django/src/rdwatch/views/tile.py

```diff
@@ -184,6 +185,79 @@ def vector_tile(
     )
 
 
+@cache_page(60 * 60 * 24 * 365)
+def satelliteimage_raster_bbox(
+    request: HttpRequest,
+):
+    if (
+        'constellation' not in request.GET
+        or 'bbox' not in request.GET
+        or 'timestamp' not in request.GET
+        or 'format' not in request.GET
+    ):
+        return HttpResponseBadRequest()
+
+    constellation = Constellation(slug=request.GET['constellation'])
+    timestamp = datetime.fromisoformat(str(request.GET['timestamp']))
+    bbox_strings = request.GET['bbox'].split(',')
+    if len(bbox_strings) != 4:
+        return HttpResponseBadRequest()
+    bbox = (
+        float(bbox_strings[0]),
+        float(bbox_strings[1]),
+        float(bbox_strings[2]),
+        float(bbox_strings[3]),
+    )
+
+    format = str(request.GET['format'])
+    # Convert generator to list so we can iterate over it multiple times
+    bands = list(get_bands(constellation, timestamp, bbox))
+
+    # Filter bands by requested processing level and spectrum
+    bands = [
+        band
+        for band in bands
+        if (band.level.slug, band.spectrum.slug)
+        == (request.GET['level'], request.GET['spectrum'])
+    ]
+
+    if not bands:
+        return HttpResponseNotFound()
+
+    # Get timestamp closest to the requested timestamp
+    precise_timestamp = min(
+        bands, key=lambda band: abs(band.timestamp - timestamp)
+    ).timestamp
+
+    # Filter out any bands that don't have that timestamp
+    bands = [band for band in bands if band.timestamp == precise_timestamp]
+
+    # Sort bands so that bands in TIF format come first (TIFs are cheaper to tile
+    # and are preferred over other formats when possible)
+    bands.sort(key=lambda band: band.uri.lower().endswith('.tif'), reverse=True)
+
+    # If the timestamp provided by the user is *exactly* the timestamp of the retrieved
+    # band, return the raster data for it. Otherwise, redirect back to this same view
+    # with the exact timestamp as a parameter so that this behavior is triggered during
+    # that request. This is done to facilitate caching of the raster data.
+    if precise_timestamp == timestamp:
+        tile = get_raster_bbox(bands[0].uri, bbox, format)
+        return HttpResponse(
+            tile,
+            content_type=f'image/{format}',
+            status=200,
+        )
+
+    query_params = {
+        **request.GET.dict(),
+        'timestamp': datetime.isoformat(precise_timestamp),
+    }
+
+    return HttpResponsePermanentRedirect(
+        reverse('satellite-bbox') + f'?{urlencode(query_params)}'
+    )
```

Comment:
I think at the minimum it would be good to factor out the common code between this endpoint and the original `z/x/y/` endpoint into a common function, if possible. I'm less sure about this, but maybe a more ideal solution would be to remove the `z/x/y/` endpoint, and have a single endpoint that takes as query parameters {`z`, `x`, `y`} OR {`bbox` }. And for now we can make the `z/x/y` redirect to the new one until the frontend is updated accordingly. But I'm conflicted on whether that is better, since the `z/x/y/` url pattern is common in GIS systems I believe. Any thoughts on this? In general, I fear that we're adding too many similar endpoints for satellite images and the API is getting bloated.

Reply:
I don't know if we can convince maplibre to use query parameters instead of paths when generating URLS.  If we do it may be a non-standard thing that people may find awkward as you mentioned.  I'd have to look into that to see if it is possible.  I do agree with refactoring the commonality between the endpoints if possible.

Reply:
Also, I'm not a fan of the current vocab of the endpoints.  'visual' means worldview is quite a leap, but I didn't want to alter the existing stuff so I used the same pattern.

Reply:
Agreed on the endpoint names, I think that's worth revisting once all of the demo code is reviewed and merged.